# Project: Wrangling and Analyze Data

## Data Gathering
In the cell below, gather **all** three pieces of data for this project and load them in the notebook. **Note:** the methods required to gather each data are different.
1. Directly download the WeRateDogs Twitter archive data (twitter_archive_enhanced.csv)

In [1]:
import pandas as pd
import numpy as np
df1 = pd.read_csv('twitter-archive-enhanced.csv')

2. Use the Requests library to download the tweet image prediction (image_predictions.tsv)

In [2]:
# Import predicted breeds file
df2 = pd.read_csv("image_predictions.tsv", sep='\t')

3. Use the Tweepy library to query additional data via the Twitter API (tweet_json.txt)

In [3]:
import json
df3_list = []

for line in open('tweet_json.txt','r'):
    data = json.loads(line)
    df3_list.append({'tweet_id': data['id'], 'timestamp': data['created_at'], 'retweets': data['retweet_count'], 
                      'favorites':data['favorite_count'], 'followers': data['user']['followers_count']})

In [4]:
df2.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [5]:
df3 = pd.DataFrame(df3_list)
df3.head()

tweet_id                       timestamp  retweets  favorites  \
0  892420643555336193  Tue Aug 01 16:23:56 +0000 2017      7168      34467   
1  892177421306343426  Tue Aug 01 00:17:27 +0000 2017      5388      29883   
2  891815181378084864  Mon Jul 31 00:18:03 +0000 2017      3553      22497   
3  891689557279858688  Sun Jul 30 15:58:51 +0000 2017      7383      37667   
4  891327558926688256  Sat Jul 29 16:00:24 +0000 2017      7926      35979   

   followers  
0    9058333  
1    9058333  
2    9058333  
3    9058333  
4    9058333

## Assessing Data
In this section, detect and document at least **eight (8) quality issues and two (2) tidiness issue**. You must use **both** visual assessment
programmatic assessement to assess the data.

**Note:** pay attention to the following key points when you access the data.

* You only want original ratings (no retweets) that have images. Though there are 5000+ tweets in the dataset, not all are dog ratings and some are retweets.
* Assessing and cleaning the entire dataset completely would require a lot of time, and is not necessary to practice and demonstrate your skills in data wrangling. Therefore, the requirements of this project are only to assess and clean at least 8 quality issues and at least 2 tidiness issues in this dataset.
* The fact that the rating numerators are greater than the denominators does not need to be cleaned. This [unique rating system](http://knowyourmeme.com/memes/theyre-good-dogs-brent) is a big part of the popularity of WeRateDogs.
* You do not need to gather the tweets beyond August 1st, 2017. You can, but note that you won't be able to gather the image predictions for these tweets since you don't have access to the algorithm used.



In [6]:
# Make copies of original pieces of data
dfarchive = df1.copy()
dfimage = df2.copy()
dftweet = df3.copy()

#### Checking Twitter Archive DataFrame

In [7]:
dfarchive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

##### Quality Issue #1: Retweets and Replies

In [8]:
dfarchive.info()
#checking info to see null values. retweeted_status_id has many null values, all non-null values are
#indications of retweets

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

##### Quality Issue #2: Checking Numerator and Denominator Columns

In [9]:
dfarchive.rating_denominator.describe()

count    2356.000000
mean       10.455433
std         6.745237
min         0.000000
25%        10.000000
50%        10.000000
75%        10.000000
max       170.000000
Name: rating_denominator, dtype: float64

In [10]:
#setting column width so that I can see all of the text
pd.set_option('display.max_colwidth', None)

In [11]:
#checking denominators less than 10

dfarchive.query('rating_denominator < 10').head()
#ID 313 is set to 960/0 but its supposed to be 13/10
#ID 516 is set to 24/7 but there is actually no rating --> remove
#ID 2335 is set to 1/2 but should be 9/10

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313   835246439529840640           8.352460e+17           26259576.0   
516   810984652412424192                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   

                      timestamp  \
313   2017-02-24 21:54:03 +0000   
516   2016-12-19 23:06:23 +0000   
2335  2015-11-16 16:11:11 +0000   

                                                                                  source  \
313   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
516   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2335  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                           text  \
313                                                                          @jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho   
516   Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx   
2335                                   This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv   

      retweeted_status_id  retweeted_status_user_id  \
313                   NaN                       NaN   
516                   NaN                       NaN   
2335                  NaN                       NaN   

     retweeted_status_timestamp  \
313                         NaN   
516                         NaN   
2335                        NaN   

                                                                                            expanded_urls  \
313                                                                                                   NaN   
516   https://www.gofundme.com/sams-smile,https://twitter.com/dog_rates/status/810984652412424192/photo/1   
2335                                      https://twitter.com/dog_rates/status/666287406224695296/photo/1   

      rating_numerator  rating_denominator  name doggo floofer pupper puppo  
313                960                   0  None  None    None   None  None  
516                 24                   7   Sam  None    None   None  None  
2335                 1                   2    an  None    None   None  None

In [12]:
#I visually QC'ed denominators greater than 10 in Microsoft Excel, but here are the text extractions 
dfarchive.query('rating_denominator > 10').text

#11/15 is actually the date, not the rating
#4/20 is the date, not the rating
#20/16 may or may not be a real rating, unclear. 


342                                                                                                               @docmisterio account started on 11/15/15
433                                                    The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd
784           RT @dog_rates: After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https:/…
902                                                                         Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE
1068          After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ
1120                             Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv
1165                                                                  

In [13]:
dfarchive.rating_numerator.describe()

count    2356.000000
mean       13.126486
std        45.876648
min         0.000000
25%        10.000000
50%        11.000000
75%        12.000000
max      1776.000000
Name: rating_numerator, dtype: float64

In [14]:
#checking to see if the rating numerators of zero are real (they are real)
dfarchive.query('rating_numerator < 1').text

315           When you're so blinded by your systematic plagiarism that you forget what day it is. 0/10 https://t.co/YbEJPkg4Ag
1016    PUPDATE: can't see any. Even if I could, I couldn't reach them to pet. 0/10 much disappointment https://t.co/c7WXaB2nqX
Name: text, dtype: object

In [15]:
#I visually QC'ed denominators greater than 10 in Microsoft Excel, but here are the text extractions 
dfarchive.query('rating_numerator < 10').text
#45 should be 13.5 not 5


45                                This is Bella. She hopes her smile made you smile. If not, she is also offering you her favorite monkey. 13.5/10 https://t.co/qjrljjt948
229     This is Jerry. He's doing a distinguished tongue slip. Slightly patronizing tbh. You think you're better than us, Jerry? 6/10 hold me back https://t.co/DkOBbwulw1
315                                                      When you're so blinded by your systematic plagiarism that you forget what day it is. 0/10 https://t.co/YbEJPkg4Ag
387                                                                                                  I was going to do 007/10, but the joke wasn't worth the &lt;10 rating
462                           RT @dog_rates: Meet Herschel. He's slightly bigger than ur average pupper. Looks lonely. Could probably ride 7/10 would totally pet https:/…
                                                                                       ...                                                       

In [16]:
#I visually QC'ed denominators greater than 10 in Microsoft Excel, but here are the text extractions 
dfarchive.query('rating_numerator > 14').text
#1712 should be 11 not 26

55                                                                                    @roushfenway These are good dogs but 17/10 is an emotional impulse rating. More like 13/10s
188                                                                                        @dhmontgomery We also gave snoop dogg a 420/10 but I think that predated your research
189                                         @s8n You tried very hard to portray this good boy as not so good, but you have ultimately failed. His goodness shines through. 666/10
285                                                                               RT @KibaDva: I collected all the good dogs!! 15/10 @dog_rates #GoodDogs https://t.co/6UCGFczlOI
290                                                                                                                                                            @markhoppus 182/10
291                                                                                                           

#### Quality Issue #3: Blank or incorrect Dog Names

In [17]:
namelist = dfarchive.name.value_counts()

In [18]:
pd.set_option('display.max_rows', None)

namelist

#581 dogs have no name, and 54 have a name called "a"

None              745
a                  55
Charlie            12
Oliver             11
Cooper             11
Lucy               11
Tucker             10
Lola               10
Penny              10
Bo                  9
Winston             9
Sadie               8
the                 8
an                  7
Bailey              7
Buddy               7
Toby                7
Daisy               7
Scout               6
Milo                6
Leo                 6
Rusty               6
Jack                6
Bella               6
Oscar               6
Stanley             6
Koda                6
Dave                6
Jax                 6
very                5
Gus                 5
George              5
Phil                5
Bentley             5
Alfie               5
Sunny               5
Louis               5
Oakley              5
Chester             5
Sammy               5
Finn                5
Larry               5
Walter              4
Winnie              4
Derek               4
Maggie    

In [19]:
dfarchive[dfarchive.name == 'a']
#a large number of these are not dogs. 
#Perhaps we should delete animals that aren't dogs using the twitter image archive,
# then come back and see if these dogs still exist in the dataframe

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
56    881536004380872706                    NaN                  NaN   
649   792913359805018113                    NaN                  NaN   
801   772581559778025472                    NaN                  NaN   
1002  747885874273214464                    NaN                  NaN   
1004  747816857231626240                    NaN                  NaN   
1017  746872823977771008                    NaN                  NaN   
1049  743222593470234624                    NaN                  NaN   
1193  717537687239008257                    NaN                  NaN   
1207  715733265223708672                    NaN                  NaN   
1340  704859558691414016                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1361  703079050210877440                    NaN                  NaN   
1368  702539513671897089                    NaN                  NaN   
1382  700864154249383937                    NaN                  NaN   
1499  692187005137076224                    NaN                  NaN   
1737  679530280114372609                    NaN                  NaN   
1785  677644091929329666                    NaN                  NaN   
1853  675706639471788032                    NaN                  NaN   
1854  675534494439489536                    NaN                  NaN   
1877  675109292475830276                    NaN                  NaN   
1878  675047298674663426                    NaN                  NaN   
1923  674082852460433408                    NaN                  NaN   
1941  673715861853720576                    NaN                  NaN   
1955  673636718965334016                    NaN                  NaN   
1994  672604026190569472                    NaN                  NaN   
2034  671743150407421952                    NaN                  NaN   
2066  671147085991960577                    NaN                  NaN   
2116  670427002554466305                    NaN                  NaN   
2125  670361874861563904                    NaN                  NaN   
2128  670303360680108032                    NaN                  NaN   
2146  669923323644657664                    NaN                  NaN   
2153  669661792646373376                    NaN                  NaN   
2161  669564461267722241                    NaN                  NaN   
2191  668955713004314625                    NaN                  NaN   
2198  668815180734689280                    NaN                  NaN   
2211  668614819948453888                    NaN                  NaN   
2218  668507509523615744                    NaN                  NaN   
2222  668466899341221888                    NaN                  NaN   
2235  668171859951755264                    NaN                  NaN   
2249  667861340749471744                    NaN                  NaN   
2255  667773195014021121                    NaN                  NaN   
2264  667538891197542400                    NaN                  NaN   
2273  667470559035432960                    NaN                  NaN   
2287  667177989038297088                    NaN                  NaN   
2304  666983947667116034                    NaN                  NaN   
2311  666781792255496192                    NaN                  NaN   
2314  666701168228331520                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2350  666050758794694657                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN         

In [20]:
dfarchive[dfarchive.name == 'None']

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
5     891087950875897856                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
30    886267009285017600           8.862664e+17         2.281182e+09   
32    886054160059072513                    NaN                  NaN   
35    885518971528720385                    NaN                  NaN   
37    885167619883638784                    NaN                  NaN   
41    884441805382717440                    NaN                  NaN   
42    884247878851493888                    NaN                  NaN   
47    883117836046086144                    NaN                  NaN   
55    881633300179243008           8.816070e+17         4.738443e+07   
59    880872448815771648                    NaN                  NaN   
62    880095782870896641                    NaN                  NaN   
64    879674319642796034           8.795538e+17         3.105441e+09   
72    878604707211726852                    NaN                  NaN   
78    877611172832227328                    NaN                  NaN   
83    876537666061221889                    NaN                  NaN   
88    875097192612077568                    NaN                  NaN   
89    875021211251597312                    NaN                  NaN   
93    874057562936811520                    NaN                  NaN   
96    873580283840344065                    NaN                  NaN   
99    872967104147763200                    NaN                  NaN   
100   872820683541237760                    NaN                  NaN   
101   872668790621863937                    NaN                  NaN   
103   872486979161796608                    NaN                  NaN   
110   871102520638267392                    NaN                  NaN   
112   870804317367881728                    NaN                  NaN   
113   870726314365509632           8.707262e+17         1.648776e+07   
125   868622495443632128                    NaN                  NaN   
127   867900495410671616                    NaN                  NaN   
130   867072653475098625                    NaN                  NaN   
131   867051520902168576                    NaN                  NaN   
133   866720684873056260                    NaN                  NaN   
137   866094527597207552                    NaN                  NaN   
141   864873206498414592                    NaN                  NaN   
148   863427515083354112           8.634256e+17         7.759620e+07   
149   863079547188785154           6.671522e+17         4.196984e+09   
154   862096992088072192                    NaN                  NaN   
155   861769973181624320                    NaN                  NaN   
157   861288531465048066                    NaN                  NaN   
160   860924035999428608                    NaN                  NaN   
164   860184849394610176                    NaN                  NaN   
165   860177593139703809                    NaN                  NaN   
168   859607811541651456                    NaN                  NaN   
170   859074603037188101                    NaN                  NaN   
172   858843525470990336                    NaN                  NaN   
177   857393404942143489                    NaN                  NaN   
179   857214891891077121           8.571567e+17         1.806710e+08   
180   857062103051644929                    NaN                  NaN   
184   856526610513747968           8.558181e+17         4.196984e+09   
185   856330835276025856                    NaN                  NaN   
186   856288084350160898           8.562860e+17         2.792810e+08   
188   855862651834028034           8.558616e+17         

##### Tidiness Issue #1: Doggo, Floofer, Pupper and Pupp are untidy columns that should be combined

In [21]:
dfarchive.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator',
       'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo'],
      dtype='object')

In [22]:
dfarchive.doggo

0        None
1        None
2        None
3        None
4        None
5        None
6        None
7        None
8        None
9       doggo
10       None
11       None
12       None
13       None
14       None
15       None
16       None
17       None
18       None
19       None
20       None
21       None
22       None
23       None
24       None
25       None
26       None
27       None
28       None
29       None
30       None
31       None
32       None
33       None
34       None
35       None
36       None
37       None
38       None
39       None
40       None
41       None
42       None
43      doggo
44       None
45       None
46       None
47       None
48       None
49       None
50       None
51       None
52       None
53       None
54       None
55       None
56       None
57       None
58       None
59       None
60       None
61       None
62       None
63       None
64       None
65       None
66       None
67       None
68       None
69       None
70       None
71    

In [23]:
dfarchive.floofer

0          None
1          None
2          None
3          None
4          None
5          None
6          None
7          None
8          None
9          None
10         None
11         None
12         None
13         None
14         None
15         None
16         None
17         None
18         None
19         None
20         None
21         None
22         None
23         None
24         None
25         None
26         None
27         None
28         None
29         None
30         None
31         None
32         None
33         None
34         None
35         None
36         None
37         None
38         None
39         None
40         None
41         None
42         None
43         None
44         None
45         None
46      floofer
47         None
48         None
49         None
50         None
51         None
52         None
53         None
54         None
55         None
56         None
57         None
58         None
59         None
60         None
61         None
62      

In [24]:
dfarchive.pupper

0         None
1         None
2         None
3         None
4         None
5         None
6         None
7         None
8         None
9         None
10        None
11        None
12        None
13        None
14        None
15        None
16        None
17        None
18        None
19        None
20        None
21        None
22        None
23        None
24        None
25        None
26        None
27        None
28        None
29      pupper
30        None
31        None
32        None
33        None
34        None
35        None
36        None
37        None
38        None
39        None
40        None
41        None
42        None
43        None
44        None
45        None
46        None
47        None
48        None
49      pupper
50        None
51        None
52        None
53        None
54        None
55        None
56      pupper
57        None
58        None
59        None
60        None
61        None
62        None
63        None
64        None
65        None
66        

In [25]:
dfarchive.puppo

0        None
1        None
2        None
3        None
4        None
5        None
6        None
7        None
8        None
9        None
10       None
11       None
12      puppo
13       None
14      puppo
15       None
16       None
17       None
18       None
19       None
20       None
21       None
22       None
23       None
24       None
25       None
26       None
27       None
28       None
29       None
30       None
31       None
32       None
33       None
34       None
35       None
36       None
37       None
38       None
39       None
40       None
41       None
42       None
43       None
44       None
45       None
46       None
47       None
48       None
49       None
50       None
51       None
52       None
53       None
54       None
55       None
56       None
57       None
58       None
59       None
60       None
61       None
62       None
63       None
64       None
65       None
66       None
67       None
68       None
69       None
70       None
71    

#### Checking Image DataFrame

In [26]:
dfimage.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [27]:
dfimage.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


##### Quality Issue #5: Identify images with false values 

In [28]:
#Here we see that the values for p1_dog are either True or False, no erroneous values here
dfimage.p1_dog.unique()

array([ True, False])

In [29]:
#here I want to see the ratio of dogs that are guessed correctly v. incorrectly. ~74% of all guesses are correct
dfimage.p1_dog.sum()/dfimage.p1_dog.count()

0.7383132530120482

In [30]:
#There are only 324 entries in which all three guesses are False
dfimage.query('p1_dog == False and p2_dog == False and p3_dog == False').tweet_id.count()

324

In [31]:
dfimage.query('p1_dog == False and p2_dog == False and p3_dog == False')

tweet_id  \
6     666051853826850816   
17    666104133288665088   
18    666268910803644416   
21    666293911632134144   
25    666362758909284353   
29    666411507551481857   
45    666786068205871104   
50    666837028449972224   
51    666983947667116034   
53    667012601033924608   
56    667065535570550784   
69    667188689915760640   
73    667369227918143488   
77    667437278097252352   
78    667443425659232256   
93    667549055577362432   
94    667550882905632768   
96    667724302356258817   
98    667766675769573376   
100   667782464991965184   
106   667866724293877760   
107   667873844930215936   
112   667911425562669056   
115   667937095915278337   
117   668142349051129856   
118   668154635664932864   
123   668226093875376128   
130   668291999406125056   
132   668466899341221888   
140   668544745690562560   
142   668614819948453888   
143   668620235289837568   
151   668643542311546881   
152   668645506898350081   
166   668981893510119424   
168   668988183816871936   
170   668992363537309700   
174   669015743032369152   
177   669214165781868544   
182   669351434509529089   
193   669571471778410496   
195   669583744538451968   
199   669661792646373376   
201   669682095984410625   
203   669749430875258880   
205   669923323644657664   
209   669972011175813120   
212   670037189829525505   
215   670055038660800512   
217   670069087419133954   
219   670079681849372674   
226   670361874861563904   
230   670408998013820928   
232   670417414769758208   
233   670420569653809152   
242   670449342516494336   
244   670465786746662913   
245   670468609693655041   
246   670474236058800128   
247   670668383499735048   
253   670727704916926465   
254   670733412878163972   
260   670783437142401025   
272   670826280409919488   
275   670838202509447168   
277   670842764863651840   
288   671159727754231808   
290   671166507850801152   
293   671347597085433856   
296   671362598324076544   
297   671390180817915904   
300   671488513339211776   
301   671497587707535361   
303   671511350426865664   
307   671533943490011136   
309   671538301157904385   
311   671544874165002241   
312   671547767500775424   
317   671744970634719232   
318   671763349865160704   
321   671855973984772097   
323   671874878652489728   
324   671879137494245376   
325   671882082306625538   
337   672231046314901505   
342   672256522047614977   
347   672475084225949696   
353   672591271085670400   
357   672609152938721280   
358   672614745925664768   
361   672828477930868736   
364   672884426393653248   
365   672898206762672129   
366   672902681409806336   
369   672970152493887488   
372   672984142909456390   
376   673148804208660480   
390   673363615379013632   
391   673576835670777856   
402   673697980713705472   
403   673700254269775872   
409   673715861853720576   
414   674008982932058114   
421   674045139690631169   
424   674063288070742018   
429   674265582246694913   
433   674318007229923329   
434   674372068062928900   
442   674632714662858753   
446   674664755118911488   
448   674690135443775488   
453   674754018082705410   
455   674767892831932416   
457   674781762103414784   
470   675113801096802304   
471   675135153782571009   
476   675153376133427200   
484   675483430902214656   
487   675501075957489664   
508   676098748976615425   
512   676215927814406144   
528   676776431406465024   
533   676897532954456065   
535   676942428000112642   
537   676948236477857792   
542   677228873407442944   
553   677573743309385728   
555   677662372920729601   
576   678675843183484930   
580   678767140346941444   
594   679503373272485890   
596   679527802031484928   
606   679877062409191424   
609   680085611152338944   
615   680176173301628928   
617   680206703334408192   
618   680221482581123072   
619   680440374763077632   
623   680583894916304897   
625   680798457301471234   
636   681242418453299201   
640   681302363064414209   
647   681679

#### Checking Twitter API DataFrame

In [32]:
dftweet.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2328 entries, 0 to 2327
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   tweet_id   2328 non-null   int64 
 1   timestamp  2328 non-null   object
 2   retweets   2328 non-null   int64 
 3   favorites  2328 non-null   int64 
 4   followers  2328 non-null   int64 
dtypes: int64(4), object(1)
memory usage: 91.1+ KB


### Quality issues
1. Remove retweets and replies

2. Remove erroneous columns regarding retweets and replies

3. Remove invalid numerator and denominator values.

4. Change numerator and denominator of ratings into a single float value. 

5. Calculate the outer fences for the dog ratings and then remove rating values that are major outliers. 

6. Replace dog names listed as "None" with NaN

7. Remove images from image dataframe that weren't guessed correctly in all three guesses. 

7.5 Remove images that are not dogs? 

8. Remove erroneous dog descriptor columns from dfarchive

### Tidiness issues
1. Variables as column headers: doggo, floofer, etc. 

2. Combine the tables into one master dataset

## Cleaning Data
In this section, clean **all** of the issues you documented while assessing. 

**Note:** Make a copy of the original data before cleaning. Cleaning includes merging individual pieces of data according to the rules of [tidy data](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html). The result should be a high-quality and tidy master pandas DataFrame (or DataFrames, if appropriate).

### Quality Issue #1:

#### Define: Remove Retweets

There are 181 retweets, so I will remove all non-null values from the column "retweeted_status_id"

In [33]:
dfarchive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

#### Code

In [34]:
#setting the dataframe equal to only the columns where retweeted_status_id and in_reply_to_status is null
dfarchive = dfarchive[dfarchive.retweeted_status_id.isnull()]
dfarchive = dfarchive[dfarchive.in_reply_to_status_id.isnull()]

#### Test

In [35]:
#checking that the dataframe now has all null values in retweeted_status_id
dfarchive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2097 non-null   int64  
 1   in_reply_to_status_id       0 non-null      float64
 2   in_reply_to_user_id         0 non-null      float64
 3   timestamp                   2097 non-null   object 
 4   source                      2097 non-null   object 
 5   text                        2097 non-null   object 
 6   retweeted_status_id         0 non-null      float64
 7   retweeted_status_user_id    0 non-null      float64
 8   retweeted_status_timestamp  0 non-null      object 
 9   expanded_urls               2094 non-null   object 
 10  rating_numerator            2097 non-null   int64  
 11  rating_denominator          2097 non-null   int64  
 12  name                        2097 non-null   object 
 13  doggo                       2097 

### Quality Issue #2:

#### Define: Remove erroneous columns from dataframe

In [36]:
dfarchive = dfarchive.drop(['in_reply_to_status_id'], axis = 1)
dfarchive = dfarchive.drop(['in_reply_to_user_id'], axis = 1)
dfarchive = dfarchive.drop(['retweeted_status_id'], axis = 1)
dfarchive = dfarchive.drop(['retweeted_status_user_id'], axis = 1)
dfarchive = dfarchive.drop(['retweeted_status_timestamp'], axis = 1)

In [37]:
dfarchive.columns

Index(['tweet_id', 'timestamp', 'source', 'text', 'expanded_urls',
       'rating_numerator', 'rating_denominator', 'name', 'doggo', 'floofer',
       'pupper', 'puppo'],
      dtype='object')

#### Code

#### Test

### Quality Issue #3:

#### Define: Remove erroneous numerators and denominators. Create a new column with a single float rating. 
In the analysis phase I determined that only a small handful (3) of ratings were stripped incorrectly. So, instead of restripping the ratings, I will simply remove erroneous ones. 


In [38]:
#I will drop rows with the following denominators: 0, 2, 7, 15, 20, 16 
#I'll remove numerators equal to 5 and 26 

#### Code

In [39]:
#First checking number of numerators
print(dfarchive.rating_numerator.unique())

[  13   12   14    5   11    6   10    0   84   24   75   27    3    7
    8    9    4  165 1776  204   50   99   80   45   60   44  121   26
    2  144   88    1  420]


In [40]:
nums = [5,26]
for num in nums:
    dfarchive = dfarchive[dfarchive['rating_numerator'] != num]

#### Test

In [41]:
#Number of numerators should be down by 2
print(dfarchive.rating_numerator.nunique())

31


#### Code

In [42]:
#Now checking number of denominators
print(dfarchive.rating_denominator.nunique())

14


In [43]:
denoms = [0,2,7,15,20,16]

for denom in denoms:
    dfarchive = dfarchive[dfarchive['rating_denominator'] != denom]

#### Test

In [44]:
#Number of denominators should be down by 6
print(dfarchive.rating_denominator.nunique())

11


### Quality Issue #4:

#### Define: Remove erroneous numerators and denominators. Create a new column with a single float rating. 
Now that I've removed erroneous rating values, I will combine them into one float rating value so that the votes can be ranked


#### Code

In [45]:
dfarchive['rating'] = (dfarchive.rating_numerator.astype(float) / dfarchive.rating_denominator.astype(float))

#### Test

In [46]:
dfarchive.rating.describe()

count    2059.000000
mean        1.179337
std         4.000923
min         0.000000
25%         1.000000
50%         1.100000
75%         1.200000
max       177.600000
Name: rating, dtype: float64

### Quality Issue #5:

#### Define: Remove major outliers from ratings
I have removed all invalid ratings, but there are still some really high ratings that will skew the data, and I have been informed that I should remove all major outliers. Thus, I will calculate the major outlier value and remove values higher than this


#### Code

In [47]:
#A major outlier is 3x outside of the quartile range

#calculating the quartile range, then multiplying it by 3. 
(1.2 - 1.1)*3


0.2999999999999996

In [48]:
#calculating the appropriate maximum for the dataset
1.2 + .3

1.5

In [49]:
dfarchive = dfarchive[dfarchive.rating < 1.5]

#### Test

In [50]:
dfarchive.rating.describe()

count    2055.000000
mean        1.069808
std         0.204276
min         0.000000
25%         1.000000
50%         1.100000
75%         1.200000
max         1.400000
Name: rating, dtype: float64

### Quality Issue #6

#### Define: Replace dogs named "None" with NaN

#### Code

In [51]:
dfarchive.name.value_counts().head(5)

None       581
a           54
Charlie     11
Lucy        11
Cooper      10
Name: name, dtype: int64

In [52]:
dfarchive.name.replace('None', np.nan, inplace=True)

#### Test

In [53]:
dfarchive.name.value_counts().head(5)

a          54
Charlie    11
Lucy       11
Cooper     10
Oliver     10
Name: name, dtype: int64

### Quality Issue #7

#### Define: Remove rows from image dataframe that did not guess correctly in all three tries

#### Code

In [54]:
#this is the number of values we want in our dataframe after the cleaning step:
dfimage.tweet_id.count() - dfimage.query('p1_dog == False and p2_dog == False and p3_dog == False').tweet_id.count()

1751

In [55]:
dfimage = dfimage.query('p1_dog == True or p2_dog == True or p3_dog == True ')

#### Test

In [56]:
dfimage.tweet_id.count()

1751

### Quality Issue #7.5

#### Define: Remove rows for images that are not dogs

#### Code

In [57]:
#First I'll look at P1 = true values for non-dogs
dftrue = dfimage[dfimage.p1_dog == True]

In [58]:
dftrue.p1.unique()
#looks like all were guessed as dogs and correct

array(['Welsh_springer_spaniel', 'redbone', 'German_shepherd',
       'Rhodesian_ridgeback', 'miniature_pinscher',
       'Bernese_mountain_dog', 'chow', 'miniature_poodle',
       'golden_retriever', 'Gordon_setter', 'Walker_hound', 'pug',
       'bloodhound', 'Lhasa', 'English_setter', 'Italian_greyhound',
       'Maltese_dog', 'malamute', 'soft-coated_wheaten_terrier',
       'Chihuahua', 'black-and-tan_coonhound', 'toy_terrier',
       'Blenheim_spaniel', 'Pembroke', 'Chesapeake_Bay_retriever',
       'curly-coated_retriever', 'dalmatian', 'Ibizan_hound',
       'Border_collie', 'Labrador_retriever', 'miniature_schnauzer',
       'Airedale', 'West_Highland_white_terrier', 'toy_poodle',
       'giant_schnauzer', 'vizsla', 'Rottweiler', 'Siberian_husky',
       'papillon', 'Saint_Bernard', 'Tibetan_terrier', 'borzoi', 'beagle',
       'Yorkshire_terrier', 'Pomeranian', 'kuvasz',
       'flat-coated_retriever', 'Norwegian_elkhound', 'standard_poodle',
       'Staffordshire_bullterrier

In [59]:
#Now I'll look at P2 = true values for non-dogs
dftrue = dfimage[dfimage.p2_dog == True]

In [60]:
dftrue.p2.unique()
#looks like all second guesses that were true were dogs

array(['collie', 'miniature_pinscher', 'malinois', 'redbone',
       'Rottweiler', 'English_springer', 'Tibetan_mastiff', 'komondor',
       'Yorkshire_terrier', 'English_foxhound', 'bull_mastiff',
       'German_shepherd', 'Shih-Tzu', 'Newfoundland', 'toy_terrier',
       'toy_poodle', 'Chesapeake_Bay_retriever', 'Siberian_husky',
       'Afghan_hound', 'bloodhound', 'papillon', 'cocker_spaniel', 'chow',
       'Irish_terrier', 'beagle', 'giant_schnauzer', 'Labrador_retriever',
       'Pembroke', 'Chihuahua', 'Weimaraner', 'Brittany_spaniel',
       'standard_schnauzer', 'vizsla', 'pug', 'Italian_greyhound',
       'Samoyed', 'Pomeranian', 'miniature_poodle', 'Lakeland_terrier',
       'Irish_setter', 'malamute', 'Border_collie', 'Leonberg',
       'French_bulldog', 'golden_retriever', 'standard_poodle', 'kuvasz',
       'Cardigan', 'silky_terrier', 'English_setter', 'Pekinese', 'boxer',
       'basset', 'Bedlington_terrier', 'Shetland_sheepdog', 'Lhasa',
       'groenendael', 'Austra

In [61]:
#Now I'll look at P3 = true values for non-dogs
dftrue = dfimage[dfimage.p3_dog == True]

In [62]:
dftrue.p3.unique()
#looks like all third guesses that were true were dogs

array(['Shetland_sheepdog', 'Rhodesian_ridgeback', 'bloodhound',
       'miniature_pinscher', 'Doberman', 'Greater_Swiss_Mountain_dog',
       'golden_retriever', 'soft-coated_wheaten_terrier',
       'Labrador_retriever', 'Pekinese', 'Ibizan_hound', 'French_bulldog',
       'malinois', 'Dandie_Dinmont', 'borzoi', 'basenji',
       'miniature_poodle', 'groenendael', 'Eskimo_dog', 'briard',
       'papillon', 'flat-coated_retriever', 'Chihuahua', 'Shih-Tzu',
       'Pomeranian', 'Saluki', 'Great_Pyrenees',
       'West_Highland_white_terrier', 'collie', 'toy_poodle', 'vizsla',
       'giant_schnauzer', 'kelpie', 'Brabancon_griffon',
       'standard_poodle', 'beagle', 'Irish_water_spaniel', 'bluetick',
       'Weimaraner', 'Chesapeake_Bay_retriever',
       'black-and-tan_coonhound', 'kuvasz', 'Staffordshire_bullterrier',
       'Yorkshire_terrier', 'Lakeland_terrier', 'cocker_spaniel',
       'Australian_terrier', 'Great_Dane', 'curly-coated_retriever',
       'schipperke', 'Newfoundla

#### Test

In [63]:
#No need to code here, by removing all wrong guesses, I also vicariously removed all non-dogs. 

##### Quality Issue #8 is taken care of after tidiness issue #1

### Tidiness Issue #1

#### Define:  Combine Doggo, Floofer, and Pupper columns into one descriptor column

#### Code: 

In [64]:
#combining the descriptors into on column
dfarchive['dog_descriptor'] = dfarchive.doggo + dfarchive.floofer +dfarchive.pupper + dfarchive.puppo

In [65]:
#looking at the column, lots of 'None's are copied
dfarchive.dog_descriptor

0           NoneNoneNoneNone
1           NoneNoneNoneNone
2           NoneNoneNoneNone
3           NoneNoneNoneNone
4           NoneNoneNoneNone
5           NoneNoneNoneNone
6           NoneNoneNoneNone
7           NoneNoneNoneNone
8           NoneNoneNoneNone
9          doggoNoneNoneNone
10          NoneNoneNoneNone
11          NoneNoneNoneNone
12         NoneNoneNonepuppo
13          NoneNoneNoneNone
14         NoneNoneNonepuppo
15          NoneNoneNoneNone
16          NoneNoneNoneNone
17          NoneNoneNoneNone
18          NoneNoneNoneNone
20          NoneNoneNoneNone
21          NoneNoneNoneNone
22          NoneNoneNoneNone
23          NoneNoneNoneNone
24          NoneNoneNoneNone
25          NoneNoneNoneNone
26          NoneNoneNoneNone
27          NoneNoneNoneNone
28          NoneNoneNoneNone
29        NoneNonepupperNone
31          NoneNoneNoneNone
33          NoneNoneNoneNone
34          NoneNoneNoneNone
35          NoneNoneNoneNone
37          NoneNoneNoneNone
38          No

In [66]:
#we can use extract to remove the nones and grab the dog names, but there are a few columns that have more than one descriptor
dfarchive.dog_descriptor.unique()

array(['NoneNoneNoneNone', 'doggoNoneNoneNone', 'NoneNoneNonepuppo',
       'NoneNonepupperNone', 'NoneflooferNoneNone', 'doggoNoneNonepuppo',
       'doggoflooferNoneNone', 'doggoNonepupperNone'], dtype=object)

In [67]:
dfarchive.query('dog_descriptor == "doggoNonepupperNone"')

tweet_id                  timestamp  \
460   817777686764523521  2017-01-07 16:59:28 +0000   
531   808106460588765185  2016-12-12 00:29:28 +0000   
575   801115127852503040  2016-11-22 17:28:25 +0000   
705   785639753186217984  2016-10-11 00:34:48 +0000   
733   781308096455073793  2016-09-29 01:42:20 +0000   
889   759793422261743616  2016-07-31 16:50:42 +0000   
1063  741067306818797568  2016-06-10 00:39:48 +0000   
1113  733109485275860992  2016-05-19 01:38:16 +0000   

                                                                                  source  \
460   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
531   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
575   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
705   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
733                      <a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>   
889   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1063  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1113  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                      text  \
460   This is Dido. She's playing the lead role in "Pupper Stops to Catch Snow Before Resuming Shadow Box with Dried Apple." 13/10 (IG: didodoggo) https://t.co/m7isZrOBX7   
531                       Here we have Burke (pupper) and Dexter (doggo). Pupper wants to be exactly like doggo. Both 12/10 would pet at same time https://t.co/ANBpEYHaho   
575                       This is Bones. He's being haunted by another doggo of roughly the same size. 12/10 deep breaths pupper everything's fine https://t.co/55Dqe0SJNj   
705   This is Pinot. He's a sophisticated doggo. You can tell by the hat. Also pointier than your average pupper. Still 10/10 would pet cautiously https://t.co/f2wmLZTPHd   
733                                                                                                             Pupper butt 1, Doggo 0. Both 12/10 https://t.co/WQvcPEpH2u   
889                       Meet Maggie &amp; Lila. Maggie is the doggo, Lila is the pupper. They are sisters. Both 12/10 would pet at the same time https://t.co/MYwR4DQKll   
1063                                                                           This is just downright precious af. 12/10 for both pupper and doggo https://t.co/o5J479bZUC   
1113                                                                                            Like father (doggo), like son (pupper). Both 12/10 https://t.co/pG2inLaOda   

                                                                                                                        expanded_urls  \
460                                                                   https://twitter.com/dog_rates/status/817777686764523521/video/1   
531                                                                   https://twitter.com/dog_rates/status/808106460588765185/photo/1   
575   https://twitter.com/dog_rates/status/801115127852503040/photo/1,https://twitter.com/dog_rates/status/801115127852503040/photo/1   
705   https://twitter.com/dog_rates/status/785639753186217984/photo/1,https://twitter.com/dog_rates/status/785639753186217984/photo/1   
733                                                                                                     https://vine.co/v/5rgu2Law2ut   
889   https://twitter.com/dog_rates/status/759793422261743616/photo/1,https://twitter.com/dog_rates/status/759793422261743616/photo/1   
1063                                                                  https://twitter.com/dog_rates/status/741067306818797568/photo/1   
1113                        

In [68]:
#for descriptors that have 'None' between them, we should just manually remove the none. 
#Otherwise the extraction will pull only the first doggo and not both descriptors
dfarchive.dog_descriptor.replace('doggoNoneNonepuppo', 'doggopuppo', inplace=True)
dfarchive.dog_descriptor.replace('doggoNonepupperNone', 'doggopupper', inplace=True)

In [69]:
# now we are extracting the descriptors, both indicvidual and combined, leaving all the None's behind
dfarchive.dog_descriptor = dfarchive.dog_descriptor.str.extract('(doggopuppo|doggopupper|doggofloofer|doggo|floofer|pupper|puppo)', expand = True)

In [70]:
dfarchive.dog_descriptor.unique()

array([nan, 'doggo', 'puppo', 'pupper', 'floofer', 'doggopuppo',
       'doggofloofer', 'doggopupper'], dtype=object)

In [71]:
#now, change the desciptors to be more gramatically correct
dfarchive.loc[dfarchive.dog_descriptor == 'doggopupper', 'dog_descriptor'] = 'doggo, pupper'
dfarchive.loc[dfarchive.dog_descriptor == 'doggopuppo', 'dog_descriptor'] = 'doggo, puppo'
dfarchive.loc[dfarchive.dog_descriptor == 'doggofloofer', 'dog_descriptor'] = 'doggo, floofer'

#### Test:

In [72]:
dfarchive.dog_descriptor.unique()

array([nan, 'doggo', 'puppo', 'pupper', 'floofer', 'doggo, puppo',
       'doggo, floofer', 'doggo, pupper'], dtype=object)

### Quality Issue #8

#### Define: Remove erroneous dog descriptor columns from table

#### Code:

In [73]:
dfarchive = dfarchive.drop(['doggo'], axis = 1)
dfarchive = dfarchive.drop(['floofer'], axis = 1)
dfarchive = dfarchive.drop(['puppo'], axis = 1)
dfarchive = dfarchive.drop(['pupper'], axis = 1)

#### Test:

In [74]:
dfarchive.columns

Index(['tweet_id', 'timestamp', 'source', 'text', 'expanded_urls',
       'rating_numerator', 'rating_denominator', 'name', 'rating',
       'dog_descriptor'],
      dtype='object')

### Tidiness Issue #2

This is where i need the most help. The tweet IDs from the image_predictions.tsv, which is downloaded from the requests library, don't match the other two dataframe IDs. 


#### Define: Combining Tables into one table

In [75]:
dfarchive.columns

Index(['tweet_id', 'timestamp', 'source', 'text', 'expanded_urls',
       'rating_numerator', 'rating_denominator', 'name', 'rating',
       'dog_descriptor'],
      dtype='object')

In [76]:
dfimage.columns

Index(['tweet_id', 'jpg_url', 'img_num', 'p1', 'p1_conf', 'p1_dog', 'p2',
       'p2_conf', 'p2_dog', 'p3', 'p3_conf', 'p3_dog'],
      dtype='object')

In [77]:
dftweet.columns

Index(['tweet_id', 'timestamp', 'retweets', 'favorites', 'followers'], dtype='object')

In [78]:
#checking to see which tweet ids from the tweet API and from the original archive dont match. 
#only 8 values dont match
set(dfarchive.tweet_id) - set(dftweet.tweet_id)

{680055455951884288,
 754011816964026368,
 759923798737051648,
 779123168116150273,
 829374341691346946,
 837366284874571778,
 844704788403113984,
 872261713294495745}

In [79]:
#checking to see which tweet ids from the tweet API and from the image prediction archive don't match. 
#it looks like no values match
set(dfarchive.tweet_id) - set(dfimage.tweet_id)

{666051853826850816,
 666104133288665088,
 666268910803644416,
 666293911632134144,
 666362758909284353,
 666411507551481857,
 666786068205871104,
 666837028449972224,
 666983947667116034,
 667012601033924608,
 667065535570550784,
 667188689915760640,
 667369227918143488,
 667437278097252352,
 667443425659232256,
 667549055577362432,
 667724302356258817,
 667766675769573376,
 667782464991965184,
 667866724293877760,
 667873844930215936,
 667937095915278337,
 668142349051129856,
 668154635664932864,
 668226093875376128,
 668291999406125056,
 668466899341221888,
 668544745690562560,
 668587383441514497,
 668614819948453888,
 668620235289837568,
 668643542311546881,
 668645506898350081,
 668981893510119424,
 668988183816871936,
 668992363537309700,
 669015743032369152,
 669214165781868544,
 669351434509529089,
 669571471778410496,
 669583744538451968,
 669682095984410625,
 669749430875258880,
 669923323644657664,
 669972011175813120,
 670055038660800512,
 670079681849372674,
 670361874861

In [80]:
#checking to see which tweet ids from the original archive and from the image prediction archive don't match. 
#it looks like no values match
#something must be wrong with the image prediction file's ID's
set(dftweet.tweet_id) - set(dfimage.tweet_id)

{666051853826850816,
 666104133288665088,
 666268910803644416,
 666293911632134144,
 666362758909284353,
 666411507551481857,
 666786068205871104,
 666837028449972224,
 666983947667116034,
 667012601033924608,
 667065535570550784,
 667070482143944705,
 667188689915760640,
 667369227918143488,
 667437278097252352,
 667443425659232256,
 667549055577362432,
 667550882905632768,
 667724302356258817,
 667766675769573376,
 667782464991965184,
 667866724293877760,
 667873844930215936,
 667911425562669056,
 667937095915278337,
 668142349051129856,
 668154635664932864,
 668226093875376128,
 668291999406125056,
 668466899341221888,
 668544745690562560,
 668587383441514497,
 668614819948453888,
 668620235289837568,
 668643542311546881,
 668645506898350081,
 668967877119254528,
 668981893510119424,
 668988183816871936,
 668992363537309700,
 669015743032369152,
 669214165781868544,
 669351434509529089,
 669571471778410496,
 669583744538451968,
 669661792646373376,
 669682095984410625,
 669684865554

In [82]:
#checking data types
#it appears all IDs have the same datatype
type(dfarchive.tweet_id[1])

numpy.int64

In [83]:
type(dftweet.tweet_id[1])

numpy.int64

In [84]:
type(dfimage.tweet_id[1])

numpy.int64

In [85]:
#proceeding with the combination of the archive data and the twitter API, since i didn't have issues with those files
df_comb = pd.merge(dfarchive,dftweet, how = 'left', on = ['tweet_id'])

In [86]:
df_comb.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2055 entries, 0 to 2054
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   tweet_id            2055 non-null   int64  
 1   timestamp_x         2055 non-null   object 
 2   source              2055 non-null   object 
 3   text                2055 non-null   object 
 4   expanded_urls       2052 non-null   object 
 5   rating_numerator    2055 non-null   int64  
 6   rating_denominator  2055 non-null   int64  
 7   name                1474 non-null   object 
 8   rating              2055 non-null   float64
 9   dog_descriptor      333 non-null    object 
 10  timestamp_y         2047 non-null   object 
 11  retweets            2047 non-null   float64
 12  favorites           2047 non-null   float64
 13  followers           2047 non-null   float64
dtypes: float64(4), int64(3), object(7)
memory usage: 240.8+ KB


In [89]:
df_comb.head()

tweet_id                timestamp_x  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   
3  891689557279858688  2017-07-30 15:58:51 +0000   
4  891327558926688256  2017-07-29 16:00:24 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0                                                       This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2                   This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB   
3                                                             This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ   
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   

                                                                                                                     expanded_urls  \
0                                                                  https://twitter.com/dog_rates/status/892420643555336193/photo/1   
1                                                                  https://twitter.com/dog_rates/status/892177421306343426/photo/1   
2                                                                  https://twitter.com/dog_rates/status/891815181378084864/photo/1   
3                                                                  https://twitter.com/dog_rates/status/891689557279858688/photo/1   
4  https://twitter.com/dog_rates/status/891327558926688256/photo/1,https://twitter.com/dog_rates/status/891327558926688256/photo/1   

   rating_numerator  rating_denominator      name  rating dog_descriptor  \
0                13                  10   Phineas     1.3            NaN   
1                13                  10     Tilly     1.3            NaN   
2                12                  10    Archie     1.2            NaN   
3                13                  10     Darla     1.3            NaN   
4                12                  10  Franklin     1.2            NaN   

                      timestamp_y  retweets  favorites  followers  
0  Tue Aug 01 16:23:56 +0000 2017    7168.0    34467.0  9058333.0  
1  Tue Aug 01 00:17:27 +0000 2017    5388.0    29883.0  9058333.0  
2  Mon Jul 31 00:18:03 +0000 2017    3553.0    22497.0  9058333.0  
3  Sun Jul 30 15:58:51 +0000 2017    7383.0    37667.0  9058333.0  
4  Sat Jul 29 16:00:24 +0000 2017    7926.0    35979.0  9058333.0

#### Code: 

#### Test:

## Storing Data
Save gathered, assessed, and cleaned master dataset to a CSV file named "twitter_archive_master.csv".

## Analyzing and Visualizing Data
In this section, analyze and visualize your wrangled data. You must produce at least **three (3) insights and one (1) visualization.**

### Insights:
1. Most favorited and retweeted tweets? 
2. Relationship between followers and retweets?
2. Most common dog breeds? 

### Visualization